From e11bfd51076e180eb10af32bf21fd29fb288cc0a Mon Sep 17 00:00:00 2001
From: Raghuveer KadarlaX <raghuveerx.kadarla@intel.com>
Date: Tue, 21 Dec 2021 20:45:17 +0800
Subject: [PATCH 101/109] misc: xpcie: fix issues with driver load unload

Fix issue with driver load/unload issue at device plug-n-play.

Signed-off-by: Raghuveer KadarlaX <raghuveerx.kadarla@intel.com>
---
 drivers/misc/xlink-pcie/remote_host/main.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/misc/xlink-pcie/remote_host/main.c b/drivers/misc/xlink-pcie/remote_host/main.c
index 205619a0b66e..b7bc040b488f 100644
--- a/drivers/misc/xlink-pcie/remote_host/main.c
+++ b/drivers/misc/xlink-pcie/remote_host/main.c
@@ -15,6 +15,8 @@
 #define HW_ID_LO_MASK	GENMASK(7, 0)
 #define HW_ID_HI_MASK	GENMASK(15, 8)
 
+static bool driver_unload;
+
 static const struct pci_device_id xpcie_pci_table[] = {
 	{ PCI_DEVICE(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_KEEMBAY), 0 },
 	{ PCI_DEVICE(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_THB_FULL), 0 },
@@ -160,8 +162,10 @@ static void intel_xpcie_remove(struct pci_dev *pdev)
 	if (xdev) {
 		intel_xpcie_pci_cleanup(xdev);
 		intel_xpcie_pci_notify_event(xdev, NOTIFY_DEVICE_DISCONNECTED);
-		intel_xpcie_list_del_device(&xdev->xpcie);
-		intel_xpcie_remove_device(xdev);
+		if (driver_unload) {
+			intel_xpcie_list_del_device(&xdev->xpcie);
+			intel_xpcie_remove_device(xdev);
+		}
 	}
 }
 
@@ -182,6 +186,7 @@ static int __init intel_xpcie_init_module(void)
 
 static void __exit intel_xpcie_exit_module(void)
 {
+	driver_unload = true;
 	pci_unregister_driver(&xpcie_driver);
 }
 
-- 
2.25.1

